administration/src/util/qrcode.ts

```diff
@@ -1,37 +1,184 @@
+  const numEcBytes = ecBlocks.getTotalECCodewords()
+  // getNumDataBytes = 196 - 130 = 66
+  const numDataBytes = numBytes - numEcBytes
+  const totalInputBytes = (calculateBitsNeeded(mode, headerBitsSize, dataBitsSize, version) + 7) / 8
+  return numDataBytes >= totalInputBytes
+}
+
+function chooseMaskPattern(
+  bits: BitArray,
+  ecLevel: QRCodeDecoderErrorCorrectionLevel,
+  version: QRCodeVersion,
+  matrix: QRCodeByteMatrix
+): number {
+  let minPenalty = Number.MAX_SAFE_INTEGER // Lower penalty is better.
+  let bestMaskPattern = -1
+  // We try all mask patterns to choose the best one.
+  for (let maskPattern = 0; maskPattern < QRCode.NUM_MASK_PATTERNS; maskPattern++) {
+    MatrixUtil.buildMatrix(bits, ecLevel, version, maskPattern, matrix)
+    let penalty = calculateMaskPenalty(matrix)
+    if (penalty < minPenalty) {
+      minPenalty = penalty
+      bestMaskPattern = maskPattern
+    }
+  }
+  return bestMaskPattern
+}
+
+// The mask penalty calculation is complicated.  See Table 21 of JISX0510:2004 (p.45) for details.
+// Basically it applies four rules and summate all penalties.
+function calculateMaskPenalty(matrix: QRCodeByteMatrix): number {
+  return (
+    MaskUtil.applyMaskPenaltyRule1(matrix) +
+    MaskUtil.applyMaskPenaltyRule2(matrix) +
+    MaskUtil.applyMaskPenaltyRule3(matrix) +
+    MaskUtil.applyMaskPenaltyRule4(matrix)
+  )
+}
+
+function createHeader(mode: QRCodeMode) {
+  // This will store the header information, like mode and
+  // length, as well as "header" segments like an ECI segment.
+  const headerBits = new BitArray()
+
+  // Do not append ECI segment, because we do not define the encoding of the QR code
+
+  // (With ECI in place,) Write the mode marker
+  QRCodeEncoder.appendModeInfo(mode, headerBits)
+  return headerBits
+}
+
+export function isContentLengthValid(content: Uint8Array): boolean {
+  let mode = QRCodeMode.BYTE
+  return willFit(mode, createHeader(mode).getSize(), content.length * 8, DEFAULT_VERSION, DEFAULT_ERROR_CORRECTION)
+}
+
+/**
+ * This function has been copied and modified from https://github.com/zxing-js/library/blob/5719e939f2fc513f71627c3f37c227e26efe06c1/src/core/qrcode/encoder/Encoder.ts#L83
+ * The main changes are:
+ * * Do not include ECI segment because we do not have a text encoding (we have binary data)
+ * * Append data as binary using `appendBits`
+ *
+ * @param content binary content
+ */
+export function encodeQRCode(content: Uint8Array): QRCode {
+  // Pick an encoding mode appropriate for the content.
+  const mode: QRCodeMode = QRCodeMode.BYTE
+
+  const headerBits = createHeader(mode)
+
+  // Collect data within the main segment, separately, to count its size if needed. Don't add it to
+  // main payload yet.
+  const dataBits = new BitArray()
+  for (let i = 0, length = content.length; i !== length; i++) {
+    const b = content[i]
+    dataBits.appendBits(b, 8)
+  }
+
+  let version: QRCodeVersion = DEFAULT_VERSION
```

Comment:
Optimally, we would dynamically choose the lowest possible version, to make the code only as dense as necessary. (And maybe throw if it is higher than a maximum version).

Reply:
I think choosing a reasonable fixed version is just fine for our usecase. We can estimate the size of the QR code pretty well. 150bytes fits pretty good right now.

Else we also have to think about reasonable upper bound.

Reply:
We could have version 8 as an upper bound. But I agree, that this is an optimization which is likely unnecessary.